This is the initial deposit for `output-codahalemetrics` for Codahale Metrics.

vonnagy opened the pull request:
This will satisfy #80 when completed.

_Do not merge yet_ as this needs a bit of discussion around the resolution  for #91 which is also an issues with the `AkkaIOStatsdCounterInterface`.

This is the initial deposit for `output-metrics` for Codahale Metrics (http://metrics.codahale.com) support. It follows the same general pattern of `output-stats`.

As described in #91, I also looking for a bit of feedback on naming the metrics as Codahale is a bit more restrictive then those being sent to DataDog. An example of the current output for an actor system named _system_ and prefix named _harness_ would be as follows:

**System wide**
harness.actor-count

**System Actors**
harness.server.system.io-tcp.duration
harness.server.system.io-tcp.delivered
harness.server.system.io-tcp.delivered.bind

**User Actors**
harness.server.user.system.cluster.duration
harness.server.user.system.cluster.delivered
harness.server.user.system.cluster.delivered.leaderchanged

Commit:
This is the initial deposit for ```output-metrics``` for Codahale Metrics support. It follows the same general pattern of ```output-stats```.

Commit:
This is the initial deposit for ```output-metrics``` for Codahale Metrics support. It follows the same general pattern of ```output-stats```.

Commit:
Fixed accidental extra comma

Commit:
Fixed accidental extra comma

Commit:
Cleaned up the default naming scheme

Commit:
Cleaned up the default naming scheme

janm399 commented:
Great; I'll take a look later on today. We will soon be moving away from Travis CI to our own Jenkins to implement a better CI / CD approach. 

I'll see if I can sort out the Travis CI build errors, but we are all getting consistent successful builds on our own machines, and consistent failures on Travis CI.

janm399 commented:
What do you think about calling the module `output-codahalemetrics`? Just `...-metrics` might be ambiguous in the future.

Now, to code. 

Intuitively, it would seem that an obvious refactoring could be moving the messages that are sent to the actors (subtypes of the `MetricsStatistic`) to the `output` module so that we can reuse them. But in this case, I think that the intuition is wrong. The messages sent to the actors map to the the statsd messages exactly, but they don't map exactly to the Codahale Metrics. So, it would make more sense to define the CM messages to match the CM API, and to perform the required conversions in the implementations of the `CounterInterface`.

I was going to write that I'm not sure what benefit we get from providing two implementations of the `CounterInterface`: one that uses (for the lack of a better word here) _blocking_ execution, and one that wraps the calls in an actor that ultimately lives in another `ActorSystem`. I can't quite see what should be the optimal solution, so let's keep both. (And same applies to Datadog / statsd.)

janm399 commented on output-metrics/src/main/scala/org/eigengo/monitor/output/metrics/AkkaMetricsCounterInterface.scala:
```diff
@@ -0,0 +1,58 @@
+  // TODO: This still has an issue that is the same in the ``AkkaIOStatsdCounterInterface`` where this class is loaded
+  // on initialization of the running application's actor system. When the 'metrics' system is loaded we run into issues
+  // with threading.
+  lazy val system = ActorSystem("metrics")
```
I think there is some argument into pulling this out into a trait that we can have in the `output` module. This will allow us to have a centrally-available fix once we discover the root cause of the threading problem.

``` scala
trait IsolatedActorSystem {
  def actorSystemName: String

  // we can figure out how to solve the AS threading problem here
  lazy val system = ActorSystem(actorSystemName);

}
```

vonnagy replied:
I think that is a good idea.

janm399 commented on output-metrics/src/main/scala/org/eigengo/monitor/output/metrics/NameMarshaller.scala:
```diff
@@ -0,0 +1,71 @@
+    def cleanPath: String = tags.head.replace("akka.path:/", "").replace('/', '.').toLowerCase
+
+    val name = aspect match {
+      case "akka.actor.count" =>
```
It would be really nice to be able to use the `Aspects` object here. However, that object is a member of the `agent-akka` module, and the output module should not be aware of where the values are coming from.

Is it possible to have the name as "akka-actor-count"? In other words, relying here on just string replacement rather than "knowing" these string values?

vonnagy replied:
The default naming scheme for statistics is totally up for discussion. I provided a way for users to provide their own `NameMarshaller` so that they can name them any way they want (we will be using this feature.

In the meantime, the logic currently works with three distinct aspects:
- akka.actor.count -> This is a global value and is not related to any specific actor. We can certainly keep the name *akka.actor.count".
- akka.queue.size -> This is specific to each actor/dispatcher and thus uses a combination of the aspect name 'queue.size' and the actor's path.
- Everything else -> These are all specific to each actor so they are a combination of the aspect (minus 'akka.actor') and their path.

If we use simple string replacement, how would we determine the 'akka.queue.size' and even 'akka.actor' replacements?

I am totally open to any suggestions on this one.

vonnagy commented:
As for the _blocking_ implementation of `CounterInterface` as defined in `MetricsCounterInterface`, I only really included that since the statsd module had it as well. I am also not sure of the benefits of this so am open to remove it if we determine it is not needed.

I will rename the module and namespaces to use `codahalemetrics` instead of `metrics` as that makes it clearer.

Commit:
Removed code that was commented out

Commit:
Removed code that was commented out

Commit:
In order to make the metrics package name clearer, the module and all associated namespaces were changed from 'metrics' to 'codahalemetrics'.

e.g. From "package org.eigengo.monitor.output.metrics" to "package org.eigengo.monitor.output.codahalemetrics"

Commit:
In order to make the metrics package name clearer, the module and all associated namespaces were changed from 'metrics' to 'codahalemetrics'.

e.g. From "package org.eigengo.monitor.output.metrics" to "package org.eigengo.monitor.output.codahalemetrics"

janm399 commented:
I'm at a client today, so I'll take a look tomorrow.

janm399 commented:
I'm not forgetting you--I promise--but I've had to deal with other stuff today. Tomorrow, I'm out again, but back on Friday.

vonnagy commented:
No worries. I am out the rest of the week for Thanksgiving here in the US, but I will check in on any comments you might have.

Commit:
Removed unused parameter "refresh".

Commit:
Merge branch 'output-metrics' of https://github.com/Webtrends/monitor into output-metrics

Conflicts:
	output-codahalemetrics/src/test/resources/META-INF/monitor/output.conf

Commit:
Added initial cut of the Codahale Metrics documentation

Commit:
Merge branch 'master' into output-metrics

vonnagy commented:
@janm399 I was adding some documentation and then will wait until the PR gets reviewed.

janm399 commented:
OK. I'm back from the madness of Scala eXchange and CodeMesh, so I'll dive into some Scala over the weekend. Thanks for your contributions & patience. Let's now rock it.

janm399 commented:
I've tidied up the code and merged in version 0.3; can you now contribute to the documentation?

vonnagy commented:
Yes, I already have most of the documentation complete so hopefully be tomorrow I should have it ready to pull. Thanks.
